### PR TITLE
Implement new crafting algorithm

### PR DESCRIPTION
**Pull request by andrenete**
I wrote a better version of crafting algorithm that, I believe, *optimally* matches inventory items to crafting grid.

The algorithm can be briefly described in the following steps:
- Get info about crafting items and groups, their positions in the crafting recipe (`craft_index`)
- Get info about inventory items that can be used in the craft, their total count in inventory (`item_index`)
- Match craft items and groups with inventory items according to the recipe and also find the amount of items for every recipe cell:
    - Item matching is easy because its always 1:1 match
    - Optimal group matching requires the specific order of iteration through groups (their items counts). For that, every group items counts are first sorted in descending order and then compared "lexicographically".
After that, we iterate through sorted groups (their items counts) and for every group we take item with max possible `total_count / (times_matched + 1)` value.
Global amount of items per cell is the min of this values.
- Check if crafting grid already contains only needed items (or empty cells) and move stacks from main inventory.

As to why the order of group iteration is important, consider an example:

|                        | Inv Item 1 (count = 28)  | Inv Item 2 (count = 15) | Inv Item 3 (count = 20)
|------------------ | ----------------------------- | ------------------------------ | -----------------------------|
| Group Cell 1 | ✓                                    | ✓                                    |                                      |
| Group Cell 2 | ✓                                    |                                        | ✓                                  |

If we start with group 2, we first match item 1 for group 2 (`28 / (0 + 1) = 28`) then item 2 for group 1 (because `[28 / (1 + 1)] < [15 / (0 + 1)]`), and the best amount we can get is 15. But actually 20 is the best (group 1 - item 1, group 2 - item 3).

The really hard case is when crafting grid is not empty. In that case the algorithm still find match for every recipe cell and moves matched items only if the craft grid already have the same items in the same positions (or empty). Actually, I don't like this behavior on non-empty grid, but there are too many cases to consider to make the algorithm work correctly (so it can, for example, ignore some cells if the crafting grid already have large amount of items in them), especially when there are groups of items in the craft recipe. Also it would require an extra re-matching step.

I believe that the algorithm works correctly with any crafting recipe whether it contains groups or not, but if you've found some bug or mistake please tell about it. I hope that this PR helps!

### CODE REVIEW

**Comment by andrenete** [comment]
Oh, I just found that I implemented recipe iteration wrong. Sorry, I will fix that!

File: register.lua [hunk 1]
```diff
@@ -527,38 +467,7 @@ local function craftguide_craft(player, formname, fields)
...
-						craft_list[craft_index] = newitem
-					end
-				end
-				index = index + 1
-			end
-		end
-	until current_count == amount
-
-	player_inv:set_list("craft", craft_list)
+	craftguide_match_craft(player_inv, "main", "craft", craft, amount)
```
Comment (Panquesito7) [inline register.lua]: This will result in a crash as that function is not with `unified_inventory.` and not a function.

Comment (andrenete) [reply register.lua]: I added a `dofile(modpath.."/match_craft.lua")` line into `init.lua`. Should I also prefix this function with `unified_inventory`?

Comment (Panquesito7) [reply register.lua]: In my opinion, I would say yes. And preferably, add the new functions to `api.lua`, so, they will be global.

Comment (andrenete) [reply register.lua]: Ok, I will add the prefix. Don't know about adding them to api functions though, because I doubt they will be reused somewhere else (at least without changing some internal data formats). I think that they are more like utility functions to do one specific task.

Comment (Panquesito7) [reply register.lua]: LGTM.

Comment (Panquesito7) [reply register.lua]: LGTM. 
If possible, make the other local functions a global function (with `unified_inventory.`).

Comment (andrenete) [reply register.lua]: Ok,  done.

Comment (Panquesito7) [reply register.lua]: You can now resolve the conversation. Thank you.

**Comment by andrenete** [comment]
Found one more problem: move code works incorrectly with unstackable items and also should check if `take_count == 0` or else some items can be moved incorrectly or disappear. I will fix that. Sorry, I should've read the previous code more carefully.

**Comment by SmallJoker** [comment]
I'm interested in a saner method to fill the crafting grid. My recent changes on this code were to make it fast (finally 10x faster), while keeping it reasonably small. But as a downside it cannot spread the items well and only provides a partial copy-to-grid function. 
Please also do come execution time benchmarking to ensure it's not worse than the old code (before #123).

**Comment by andrenete** [comment]
In #123 you wrote that you test "various" recipes for crafting 10 items. Do you remember what recipes and items in inventory you tried, so we can compare the results? Maybe you could test it better if you know some worst cases?

The worst time I got so far is when crafting locked chest when having main inventory filled with different types of wood planks and steel: `3735 μs`. The average from some craft recipe - inventory items combinations is around `2000 μs`.

For testing, I measured the time before and after `unified_inventory.craftguide_match_craft` call in `register.lua`.

**Comment by andrenete** [comment]
This non empty grid case made me think: maybe it's better to forbid players to keep things in the craft grid? Because that's not what it's for anyway.
Is it possible to return items to the main inventory when the player closes the inventory page? Or when there is no space in the main inventory just drop items on the ground? This is what minecraft does and I think it's much better than to manage matching leftovers and stuff by hand.

**Comment by SmallJoker** [comment]
> Do you remember what recipes and items in inventory you tried, so we can compare the results? 

That was the bookshelf recipe (99x), also with different kinds of planks and sometimes partially filled crafting slots.

> Or when there is no space in the main inventory just drop items on the ground?

No, this behaviour may not change. Player expect the copy function to be safe since it's always been that way for years. You could move items stackwise back to the inventory, and skip those where's no space available.

3.7ms is quite good compared to the version before #123 (you might test that for comparison as well). All fine regarding the execution time.

**Comment by andrenete** [comment]
I tested the bookshelf recipe. The worst case that I've been able to get is when I tried to craft "all" but with already filled crafting grid for 10 bookshelves ([See video](https://imgur.com/a/n0i64GO)).

It turns out that the biggest time was spent on actual moving and not the matching process:

Craft index time: 44 μs
Item index time: 376 μs
Match time: 682 μs
Move check time: 13 μs
Move time: 10575 μs -- `for match_pos, item_name in pairs(matched_items) do ... end`
Inv set time: 599 μs

Total time: 12285 μs

I tried that recipe again, but this time I measured the time between `inv:remove_item(src_list_name, matched_stack)` in the moving loop:
348 μs
1902 μs
591 μs
602 μs
818 μs
900 μs
909 μs
708 μs
637 μs

Total 7415 μs

So it all comes down to the fact that `remove_item` works pretty slow.

**Comment by andrenete** [comment]
I just compiled a new minetest version from source and item removing seems to be working faster. But now there's a bug that main inventory doesn't immediately updates after stack moving ([Video](https://imgur.com/a/4MHsuM9)). It affects both my version and master version of crafting code, so I guess it's some core problem.

**Comment by SmallJoker** [comment]
> I guess it's some core problem.

It indeed is (or rather was). Fixed by now, don't worry about this issue.

**Comment by andrenete** [comment]
I finally made it work correctly with non empty crafting grid. Code is extremely ugly now, but takes many more cases into account, can swap unneeded items, swap items inside crafting grid if main inventory is full. Need lots of refactoring and also corner cases checking.

**Comment by SmallJoker** [comment]
PS: The change to optimize `remove_item` (and alike) is already pending. https://github.com/minetest/minetest/pull/8856
After it's merged, these operations will be remarkably faster (did not speed test yet).

**Comment by andrenete** [comment]
I'm still thinking about the correct implementation of the mathing algorithm and I realized that my current approach is wrong when groups in the crafting recipe overlap (inventory contains items that belong to multiple groups).

To clear things up, the task I'm trying to solve is follow:

- Suppose we have a set of resources: {r1, r2, r3, ..., rn}.
This is our items in inventory. Each resource ri is determined by its total count: ri.count.

- We also have a set of cells that we can put resources into: {c1, c2, ..., ck}.
This cells represent the crafting grid and in our case k = 9.

- For a specific crafting recipe we know, what resources we can put in every cell. In other words we have a map:
c1 -> {r11, r21, ..., rn1}
c2 -> {r21, r22, ..., rn2}
...
ck -> {r1k, r2k, ..., rnk}

The task is to select exactly one resource for every cell (if it's possible, of course), such that the minimum resource count per cell is maximal. Note that one resource can be splitted between multiple cells, but one cell can contain only one resource (because different items are unstackable).

Example:

{r1, r2, r3}
r1.count = 30
r2.count = 28
r3.count = 20

{c1, c2, c3}
c1 -> {r2, r3}
c2 -> {r1, r2}
c3 -> {r1}

Best:
c1 -> r3
c2 -> r2
c3 -> r1
Minimum count: 20 

Not best:
c1 -> r2
c2 -> r1
c3 -> r1
Minimum count: 15 (divide r1 between two cells)

For this example, my algorithm will produce the second result, which is worse than the first one. So 'cell lookup order' approach is incorrect here.

We can also imagine our map as a mathematical graph with resources and cells as vertices and 'resource can be put in cell' connections as edges. Since edges only connect resources with cells, this graph is a bipartite graph.

Maybe there are existing algorithms that already solve this?

**Comment by SmallJoker** [comment]
mind = blown

I never analysed algorithms this percise, neither am I familiar with any words mentioned here.
> Since edges only connect resources with cells, this graph is a bipartite graph.

From what I can tell: There's yet no function to count the total of similar nodes within the inventory, but that should be doable. My approach for this craft recipe copying would be the following:

1) Iterate through any crafting slot
2) For each stack: check whether there's enough matching stacks in the main list
3) Select the biggest (in total) ItemStack type that matches the required group(s)
4) If the crafting stack is non-empty: look out for that specific ItemStack
5) If there are not enough same ItemStacks: put the crafting stack back to the main list and do step 3

**Comment by andrenete** [comment]
I believe the hardest part is step 3. If we take biggest stack every time, we may end up with unfilled crafting slots, or with unoptimal placement of items in the grid. How this would work, for example, if I have just one stack 'copper 99' in main and want to craft 'All' copper blocks?

**Comment by SmallJoker** [comment]
Not the biggest stack, but the sum of all similar stack. Counting all unique types together would solve this issue. This way you'd always know how many items of each kind are in total left over.

**Comment by SmallJoker** [comment]
I meant the sum of all same stacks. Then the sum of all stacks within each group. Latter needs re-calculating after each movement, but this way you could spread the required items per craft slot proportionally.

**Comment by andrenete** [comment]
Proportionally yes, but not optimally.

Imagine you want to craft a wooden sword. In main you have 99 sticks, 4 pine wood planks and 12 apple wood planks. Wood group total = 16 and we have two crafting slots that require wood group. So we divide 16 by 2 and get 8. Then we start moving:
- max ItemStack for first slot is apple wood planks. We take 8 and decrease count: total = 8, apple wood planks = 4.
- max ItemStack for second slot is also apple wood planks (or pine because we now  have equal amount of them). We take 4 and stop.

So for this two slots we match 8 and 4. But best is 6 and 6 (12 apple wood planks / 2). Sticks count doesn't really matter here, 99 just because it's more that wood group total.

I think this is why I rejected this idea at the start.

File: match_craft.lua [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+
+		inv:set_stack(dst_list_name, i, leftover)
+	end
+
+	local skipped = {}
+
+	for match_pos = 1, dst_count do
+		local item_name = matched_items[match_pos]
+
+		if item_name ~= nil then
```
Comment (SmallJoker) [inline match_craft.lua]: You might want to put this into a pseudo loop to reduce the indents below:

```lua
repeat
	-- ...
	if src_leftover:is_empty() then
		break
	end
	-- ...
until true
```

File: match_craft.lua [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+	local craft_index = item_index.craft_index
+	local index_used = item_index.used_items
+	local index_unused = item_index.unused_items
+
+	local list_count = #inv_list
+
+	for i = 1, list_count do
+		local stack = inv_list[i]
+
+		if not stack:is_empty() then
```
Comment (SmallJoker) [inline match_craft.lua]: Maybe repeat-until here to reduce indent levels as well?

**Comment by SmallJoker** [comment]
Sorry for the delay. I just tested the PR, and it works really well - at least after the optimizations in 5.1.0-dev.

After some tests I somehow cannot get the craft grid to work any more:
```
2019-10-10 19:33:22: WARNING[Main]: GUIFormSpecMenu::drawList(): The inventory list "craft" @ "current_player" doesn't exist
```
This happens when the size is changed to 0. However, I don't see where in the code it would do that. Did a few tests to determine the source of this problem:
 * Minetest 5.1.0-dev (10 June) + minetest_game HEAD -> problem occurs
 * Minetest 5.1.0-dev (10 June) + minetest_game 0.4.17.1 -> problem occurs
 * Minetest 5.1.0-dev (15 July) + minetest_game HEAD -> problem occurs
 * Minetest 5.1.0-dev (9 Sept) + minetest_game HEAD -> problem occurs

Any ideas?

**Comment by SmallJoker** [comment]
Sorry for the delay. I just tested the PR, and it works really well - at least after the optimizations in 5.1.0-dev.

Test conditions: 4x acacia wooden planks stacks + 4x books stacks, craft "All" bookshelves
5.1.0-dev (10 June):
```
Exec time: 3608 μs
Exec time: 7116 μs -- non-empty craft grid
Exec time: 4173 μs
Exec time: 12708 μs -- non-empty craft grid
Exec time: 4456 μs
```
5.1.0-dev (9 Oct):
```
Exec time: 1005 μs
Exec time: 1205 μs
Exec time: 1126 μs -- non-empty craft grid
Exec time: 1099 μs
Exec time: 1101 μs -- non-empty craft grid
```

Would you please be so nice to add an introduction to "match_crafts.lua", to understand the logic behind this code? Also some function descriptions would be incredible helpful if any bugs show up in the future.

**Comment by andrenete** [comment]
Looks like the problem with the optimal placement is more complicated than I thought... Instead, I gave up this "lexicographical sort" idea and rewrote the entire code from scratch to make it more understandable and simple. And besides, I don't even know if there are recipes with overlapping groups of items....
I also added a description for each function.

**Comment by andrenete** [comment]
As for indents, the maximum depth has been reduced to 5. I don't think a pseudo loop is needed now.

**Comment by andrenete** [comment]
Sorry, made one small change again. Depth is 6 now, but I think its ok.

File: match_craft.lua [hunk 4]
```diff
@@ -349,16 +349,19 @@ function unified_inventory.move_match(inv, src_lists, dst_list, match_table, amo
...
 
-			if not occupied:is_empty() then
-				local leftover = unified_inventory.add_item(inv, src_lists, occupied)
-
-				if not leftover:is_empty() then
-					inv:set_stack(dst_list, pos, leftover)
-					break
+			repeat
+				if not occupied:is_empty() then
+					local leftover = unified_inventory.add_item(inv, src_lists, occupied)
```
Comment (SmallJoker) [inline match_craft.lua]: Could there be an edge-case where it cannot be added do any stack?
Also, what's the difference to an integrated loop (below)? It should do the same.
```Lua
while not occupied:is_empty() do
...
end
``

Comment (andrenete) [reply match_craft.lua]: https://imgur.com/a/5r6h1V1

Actually, the last corner case is when at the moment of matching there is no free space for `occupied` item, but there IS after the whole process of moving. I dont really think this can be fixed without "peeking" other positions at the current step, but lets stop here and not overcomplicate it even more :smiley:

**Comment by andrenete** [comment]
Just discovered that inv:remove_item allow for oversized stacks, but inv:add_item does not?

**Comment by andrenete** [comment]
I'm so sorry, current code doesn't work for recipes with unstackable items with wear or metadata. Removing all `bounded_amount * count_pos` items at once can lead to metadata loss.
But removing them sequentially in the for pos in (pos_set) loop can lead to other problems... will think about it.

**Comment by SmallJoker** [comment]
> I mean shouldn't it add stacks with at most stack_max items (99 + 99 instead of 198, for example)?

There were so many "bugfixes" in that area that I don't remember any more what exactly the side-effects of that behaviour were. The inventory mechanism is quite generic, hence rules for players also apply to mods.
I think there were also some functions that did not return the correct leftover stack sizes. Please check whether 5.1.0 works the same way (or preferably, better).

**Comment by andrenete** [comment]
I fixed the issue with metadata and placement now works for this last corner case (screenshot that I added earlier). I think this is the final version.

**Comment by andrenete** [comment]
Please do not merge yet (if you wanted to do so). There's one more problem I want to fix.

**Comment by andrenete** [comment]
Taking more items than needed to free up positions wasn't a very good idea, especially for oversized stacks (because pre-removed items can't be re-added to this stacks again) so I removed this part of the code. Better take only what we need for the recipe and not overcomplicate it for some very rare corner cases.

**Comment by andrenete** [comment]
@SmallJoker, ping?

**Comment by SmallJoker** [comment]
Will re-test this weekend and merge it afterwards.
